net/tstun/wrap.go

```diff
@@ -235,40 +260,29 @@ func (t *Wrapper) Name() (string, error) {
 // so packets may be stuck in t.outbound if t.Read called t.tdev.Read directly.
 func (t *Wrapper) poll() {
 	for {
-		select {
-		case <-t.closed:
+		if t.closed() {
+			return
+		}
+		buf := <-t.bufferC
+		// nil buffer means t is closed.
+		if buf == nil {
 			return
-		case <-t.bufferConsumed:
-			// continue
 		}
 
 		// Read may use memory in t.buffer before PacketStartOffset for mandatory headers.
 		// This is the rationale behind the tun.Wrapper.{Read,Write} interfaces
 		// and the reason t.buffer has size MaxMessageSize and not MaxContentSize.
-		n, err := t.tdev.Read(t.buffer[:], PacketStartOffset)
-		if err != nil {
-			select {
-			case <-t.closed:
-				return
-			case t.errors <- err:
-				// In principle, read errors are not fatal (but wireguard-go disagrees).
-				t.bufferConsumed <- struct{}{}
-			}
-			continue
-		}
-
+		// In principle, read errors are not fatal (but wireguard-go disagrees).
+		n, err := t.tdev.Read(buf, PacketStartOffset)
 		// Wireguard will skip an empty read,
 		// so we might as well do it here to avoid the send through t.outbound.
-		if n == 0 {
-			t.bufferConsumed <- struct{}{}
+		if n == 0 && err == nil {
+			t.bufferC <- buf
```

Comment:
Confused by how this works. poll both reads from and writes to this channel. Is it just talking to itself? Why doesn't it just hold onto the buf?

Reply:
Yes, it is talking to itself. I've added a new commit changing it to an inner loop. (Will push soon.) Do you think that that is clearer?